src/Shared/FileMatcher.cs

```diff
@@ -1953,12 +1932,44 @@ private static string ComputeFileEnumerationCacheKey(string projectDirectoryUnes
 
             using (var sb = new ReuseableStringBuilder(projectDirectoryUnescaped.Length + filespecUnescaped.Length + excludeSize))
             {
-                if (pathValidityExceptionTriggered)
+                var pathValidityExceptionTriggered = false;
+
+                try
                 {
-                    sb.Append(projectDirectoryUnescaped);
+                    // Ideally, ensure that the cache key is an absolute, normalized path so that other projects evaluating an equivalent glob can get a hit.
+                    // Corollary caveat: including the project directory when the glob is independent of it leads to cache misses
+
+                    var filespecUnescapedFullyQualified = Path.Combine(projectDirectoryUnescaped, filespecUnescaped);
+
+                    if (filespecUnescapedFullyQualified.Equals(filespecUnescaped, StringComparison.Ordinal))
+                    {
+                        // filespec is absolute, don't include the project directory path
+                        sb.Append(filespecUnescaped);
+                    }
+                    else
+                    {
+                        // filespec is not absolute, include the project directory path
+                        // differentiate fully qualified filespecs vs relative filespecs that got prepended with the project directory
+                        sb.Append("p");
```

Comment:
Can you either a) use longer, more descriptive strings or b) make consts of the short strings with descriptive names and use them?

Reply:
Consts it is. I'd rather keep these as short as possible to keep the potential hash collision lookups fast. Alternatively, we could transition to an unsigned int based hash instead of the actual string, but let's leave that until telemetry tells us this one's slow :)